Build and Release Action

jcjones opened the pull request.

jcjones committed:
Draft Build and Release CI action

jsha commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,40 @@
+  # Runs automatically when a tag beginning with 'release-' is pushed.
+  push:
+    tags:
+      - release-*
```
With Boulder we do dated releases; for unbound_exporter we do versioned releases (https://github.com/letsencrypt/unbound_exporter/releases). So let's make this trigger on tags matching `v*`.

jsha commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,40 @@
+
+          fpm -f -s dir -t deb --name "prometheus_unbound_exporter" \
+            --license "Apache 2.0" --vendor "ISRG" \
+            --url "https://github.com/letsencrypt/unbound_exporter" --prefix=/opt/prometheus/bin/ \
```
I think /opt/ is a Fedora/CentOS thing. I would expect this to go in /usr/bin/.

jsha commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,40 @@
+            --license "Apache 2.0" --vendor "ISRG" \
+            --url "https://github.com/letsencrypt/unbound_exporter" --prefix=/opt/prometheus/bin/ \
+            --version "$(GITHUB_REF)" --iteration "$(GITHUB_SHA)" --epoch "1" \
+            --package "prometheus_unbound_exporter-$(GITHUB_REF)-$(GITHUB_SHA).x86_64.deb" \
```
Let's name this after just the tag, so

```suggestion
            --package "prometheus_unbound_exporter-$(GITHUB_REF).x86_64.deb" \
```

jsha commented on tools/make-deb.sh:
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
+#
+# This script expects to run on Ubuntu. It installs the dependencies necessary
+# to build Boulder and produce a Debian Package. The actual build and packaging
```
s/Boulder/unbound_exporter/

jsha commented on tools/make-deb.sh:
```diff
@@ -0,0 +1,41 @@
+GO_VERSION=$(grep "BOULDER_TOOLS_TAG:-" docker-compose.yml | sed -e 's/.*-go//' -e 's/_.*//')
+
+# Download and unpack our production go version.
+$(dirname -- "${0}")/fetch-and-verify-go.sh "${GO_VERSION}"
```
Do we have fetch-and-verify-go in this repo?

jsha commented on tools/make-deb.sh:
```diff
@@ -0,0 +1,41 @@
+export PATH=/usr/local/go/bin:$PATH
+
+# Install fpm, this is used in our Makefile to package Boulder as a deb or rpm.
+sudo gem install --no-document fpm
```
I overlooked this in our Boulder releaser (and will revisit) but we should install a specific fpm version.

jsha replied:
https://github.com/letsencrypt/boulder/pull/6049

jsha commented on tools/make-deb.sh:
```diff
@@ -0,0 +1,41 @@
+export ARCHIVEDIR="${PWD}"
+
+# Build Boulder and produce a Debian Package at $PWD.
+make deb
```
I think we don't have a Makefile here. Let's name this something like "install-build-deps.sh" and remove these lines.

jcjones committed:
more hacking, less working

jprenken committed:
Split GitHub Actions into multiple jobs

- Replace make-deb script with nfpm usage
- Remove "prometheus_" prefix from packaging
- Move binary from /usr/local/bin to /usr/bin

jprenken committed:
Remove extra '/tmp'

jprenken committed:
Improve build step separation

jprenken committed:
Simplify YAML syntax
